lib/scholar/metrics.ex

```diff
@@ -29,34 +54,65 @@ defmodule Scholar.Metrics do
 
   ## Examples
 
-      iex> Scholar.Metrics.accuracy(Nx.tensor([[0, 1], [1, 0], [1, 0]]), Nx.tensor([[0, 1], [1, 0], [0, 1]]))
+      iex> Scholar.Metrics.onehot_accuracy(Nx.tensor([[0, 1], [1, 0], [1, 0]]), Nx.tensor([[0, 1], [1, 0], [0, 1]]))
       #Nx.Tensor<
         f32
         0.6666666865348816
       >
+      iex> y_true = Nx.tensor([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]])
+      iex> y_pred = Nx.tensor([[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 0, 1]])
+      iex> Scholar.Metrics.onehot_accuracy(y_true, y_pred)
+      #Nx.Tensor<
+        f32
+        0.5
+      >
 
   """
-  defn accuracy(y_true, y_pred) do
+  defn onehot_accuracy(y_true, y_pred) do
     assert_shape(y_true, Nx.shape(y_pred))
 
-    transform({y_true, y_pred}, fn {y_true, y_pred} ->
-      if elem(Nx.shape(y_pred), Nx.rank(y_pred) - 1) == 1 do
-        y_pred
-        |> Nx.greater(0.5)
-        |> Nx.equal(y_true)
-        |> Nx.mean()
-      else
-        y_true
-        |> Nx.argmax(axis: -1)
-        |> Nx.equal(Nx.argmax(y_pred, axis: -1))
-        |> Nx.mean()
-      end
-    end)
+    y_true
+    |> Nx.argmax(axis: -1)
+    |> Nx.equal(Nx.argmax(y_pred, axis: -1))
+    |> Nx.mean()
+  end
+
+  @doc ~S"""
+  Computes the accuracy of the given predictions
+  for multi-class classification problems.
+
+  ## Argument Shapes
+
+    * `y_true` - $\(d_0, d_1, ..., d_n\)$
+    * `y_pred` - $\(d_0, d_1, ..., d_n\)$
+
+  ## Examples
+
+      iex> y_true = Nx.tensor([0, 1, 1, 1, 1, 0, 2, 1, 0, 1], type: {:u, 32})
+      iex> y_pred = Nx.tensor([0, 2, 1, 1, 2, 2, 2, 0, 0, 1], type: {:u, 32})
+      iex> Scholar.Metrics.accuracy(y_true, y_pred, num_classes: 3)
+      #Nx.Tensor<
+        f32[3]
+        [0.6666666865348816, 1.0, 0.25]
+      >
+
+  """
+  defn accuracy(y_true, y_pred, opts \\ []) do
+    opts = keyword!(opts, [:num_classes])
+    assert_shape_pattern(y_true, {_})
+    assert_shape(y_pred, Nx.shape(y_true))
+
+    cm = confusion_matrix(y_true, y_pred, opts)
+
+    true_positive = Nx.take_diagonal(cm)
+    false_positive = Nx.subtract(Nx.sum(cm, axes: [0]), true_positive)
+
+    Nx.divide(true_positive, true_positive + false_positive + 1.0e-16)
```

Comment:
This is precision :)

Reply:
Ah, actually, for multi-class classification problems, precision = recall = accuracy = micro f1 score.<sup>*</sup>

<sup>*</sup>: when each observation/input only associates with a single label. Otherwise, it should be multi-label multi-class classification problems (which we need to handle in a different way).

Reply:
We should return a single value, no? That's what [sklearn.metrics.accuracy_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html) does.

Reply:
sorry brain not working for a moment 😂 should be fixed now